# CSgenetics scRNA-seq
**A fully automated Nextflow pipeline for the sequencing data from CS Geneticsâ€™ kits, based on foundational scientific insights and sophisticated innovations single-cell RNA-Seq data**.

[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.10.3-23aa62.svg)](https://www.nextflow.io/)
[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed?labelColor=000000&logo=docker)](https://www.docker.com/)
[![Follow on Twitter](http://img.shields.io/badge/twitter-%40csgenetics-1DA1F2?labelColor=000000&logo=twitter)](https://twitter.com/csgenetics)

## Introduction

**csgenetics/scrnaseq** is a bioinformatics best-practice analysis pipeline for processing -- single-cell RNA-seq data.

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It uses Docker containers making installation trivial and results highly reproducible.

It convert the FASTQ files generated by their sequencing experiment to counts matrices that can be loaded into Seurat or scanpy for clustering

## Getting started

### a. Generating Genome Indexes

STAR source code and binaries can be downloaded from GitHub: named releases from https://github.com/alexdobin/STAR/releases

Strongly recommended that users generate their own genome indexes with most up-to-date assemblies and annotations

Examples of acceptable genome sequence files:
- **ENSEMBL**: Download the FASTA file containing all the chromosomes together in the [Human Ensembl](https://www.ensembl.org/Homo_sapiens/Info/Index) genome, which has primary assembly in the filename. Navigate to the Gene annotation section of the Ensembl website and click on the Download [Human GTF link](https://ftp.ensembl.org/pub/release-103/gtf/homo_sapiens/). This takes you to an FTP site with a list of GTF files available.

- **GENCODE**: Files marked with PRI (primary) should be used as a refrences files. Strongly recommended for mouse and human:http://www.gencodegenes.org/

The basic options to generate genome indices using STAR are as follows:

```bash 
STAR \
    --runMode genomeGenerate \
    --genomeDir /path/to/genomeDir \
    --genomeFastaFiles /path/to/genome/fasta.fa \
    --runThreadN 4 \
    --sjdbGTFfile /path/to/annotations.gtf
```
### b. Manually  builds Docker images from a Dockerfile
Build an example image from a Dockerfile

```bash
cd images/Report/
docker build \
    --tag 613396392907.dkr.ecr.us-east-1.amazonaws.com/bio-csgenetics-report:0.0.1 . 
```
where

```bash
Change directory where dockerfile and resource files are present to build an image
--tag   Name and optionally a tag in the name:tag format
.       (dot) which is an alias for $PWD current directory     
 
```

## Pipeline-Environments

Different options are available to run the pipeline using various configurations in this repo:

### Nextflow with docker images 

The following are required to run the CSgenetics scRNAseq pipeline;
1. Prerequisite software
    - [`Nextflow`](https://nf-co.re/usage/installation) (preferably v21.x or higher)
    - [`Docker`](https://docs.docker.com)

2. Reference genome
    - Already build index files are stored for Human (GRCh38) and Mouse (GRCm39) genome in [S3 bucket](https://s3.console.aws.amazon.com/s3/buckets/bptest0?prefix=csgenetics/&region=us-east-1).
    - Please refer to Generating Genome Indexes section for custom genome downloading and building STAR index.

      **NOTE**:- Please make sure to change the species_path option in the nextflow.config based on genome used.

3. Clone the csgenetics repository by running the following command:

```bash
git clone https://github.com/basepair/csgenetics.git && cd csgenetics
```

4. Test it on a minimal dataset with a single command

```bash
nextflow run main.nf -profile test
```

5. Start running your own analysis!
    - See [usage docs](docs/usage.md) for all of the available options when running the pipeline.
   
```bash
nextflow run main.nf -profile <docker>
```
### Dockerized

The pipeline is containerized and can be run as-is with the following commands to execute the pipeline on sample data:

Prerequisites: Software [`Docker`](https://docs.docker.com) & Clone the Repository 
```bash
git clone https://github.com/basepair/csgenetics.git /home/$USER/analysis && cd /home/$USER/analysis
```


Setting the stage for running pipeline

Pulling or updating the docker container 

```bash
docker pull 613396392907.dkr.ecr.us-east-1.amazonaws.com/bio-csgenetics-scrna:0.0.1

```

```bash
docker run -it --rm -e USER=$USER \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v /home/$USER/analysis:/home/$USER/analysis \
  613396392907.dkr.ecr.us-east-1.amazonaws.com/bio-csgenetics-scrna:0.0.1 \ 
  nextflow run main.nf -profile docker -w /home/$USER/analysis/work/ --outdir /home/$USER/analysis/results/
```
where
```bash
-it --rm runs the container interactively and cleans up after exiting
-e USER=$USER makes your USER environmental variable visible inside the container
-v /var/run/docker.sock:/var/run/docker.sock enables Docker to spawn "sister" containers
-v /home/$USER/analysis:/home/$USER/analysis sets up a persistent working directory for all containers
```

- See [usage docs](docs/usage.md) for all of the available options when running the pipeline.

### Nextflow Local/Manual Installation

If the software is not already available, you will need to install it

#### Dependencies

##### Core tools

* Unix-like operating system (Linux, macOS, etc)
* [Nextflow](https://www.nextflow.io/)  >=21.10.3
* [JAVA](https://www.oracle.com/java/technologies/downloads/) >=11
* [Basespace-cli](https://developer.basespace.illumina.com/docs/content/documentation/cli/cli-overview) 0.1 

##### Bio tools

* [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) 0.11.9
* [Fastp](https://github.com/OpenGene/fastp/releases/tag/v0.23.1) 0.23.1
* [Umi-tools-csgx](https://pypi.org/project/umi-tools-csgx/) 1.1.1
* [Samtools](http://www.htslib.org/download/) 1.14
* [Star](https://github.com/alexdobin/STAR/releases/tag/2.7.9a) 2.7.9a
* [Subread](https://subread.sourceforge.net/) 2.0.1
* [Multiqc](https://multiqc.info) 1.11
* [Qualimap](https://bioconda.github.io/recipes/qualimap/README.html) 2.2.2d

##### Python packages

* [Python](https://www.python.org/downloads/) 3.9.7
* [Pysam](https://pypi.org/project/pysam/0.19.1/) 0.19.1
* [Scanpy](https://pypi.org/project/scanpy/1.7.2/) 1.7.2

##### R packages

* [R](https://www.r-project.org/) 4.2.1
* optparse 1.7.3              
* R2HTML 2.3.3               
* rjson 0.2.21                
* scales 1.2.1               
* ggplot2 3.4.1               
* DropletUtils 1.18.1        
* SingleCellExperiment 1.20.0 
* SummarizedExperiment 1.28.0
* Biobase 2.58.0             
* GenomicRanges 1.50.2       
* GenomeInfoDb 1.34.9         
* IRanges 2.32.0             
* S4Vectors 0.36.1            
* BiocGenerics 0.44.0        
* MatrixGenerics 1.10.0       
* matrixStats 0.63.0         
* Matrix 1.5-3     
  


Start running your own analysis!

- See [usage docs](docs/usage.md) for all of the available options when running the pipeline.
   
```bash
nextflow run main.nf -profile <standard>
```

