# CS Genetics scRNA-Seq pipeline
**A Nextflow pipeline for processing scRNA-Seq data generated using CS Genetics' single-cell kit to produce a genes by barcode count table.**

[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.10.3-23aa62.svg)](https://www.nextflow.io/)
[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed?labelColor=000000&logo=docker)](https://www.docker.com/)

## Contents
- [Introduction](#introduction)
- [Running the pipeline](#running-the-pipeline)
    - [Dockerized](#dockerized)
    - [Locally with dockerized processes](#locally-with-dockerized-processes)
- [Specifying input sequencing files](#specifying-input-sequencing-files)
- [Testing the pipeline](#testing-the-pipeline)
- [Launching the pipeline directly the Github repo](#launching-the-pipeline-directly-from-the-csgeneticcsgenetics_scrnaseq-github-repo)


## Introduction

**CS Genetics' scRNA-Seq pipeline** is a bioinformatics best-practice analysis pipeline for processing single-cell RNA-Seq data from their single-cell RNA-Seq kits.

It runs on a Unix-like operating system (Linux, macOS, etc)

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a portable manner. The pipeline uses Docker containers to run the main pipeline instance and its constituent processes making installation trivial and results reproducible.

It processes FASTQ files generated by CS Genetics' scRNA-Seq library kit to count matrices that can be loaded directly into [Seurat](https://satijalab.org/seurat/index.html) or [scanpy](https://scanpy.readthedocs.io/en/stable/) for further analyses.
<div style="text-align: right"><a href="#cs-genetics-scrna-seq-pipeline">top</a></div>

## Running the pipeline

The pipeline can be run in one of two ways:
- **Locally with dockerized processes (recommended)**
- **Dockerized**

### Locally with dockerized processes (recommended)
The pipeline is run natively in your environment but using docker containers for each of the pipeline's processes.

To run the pipeline locally with dockerized processes, you must have Nextflow ([installation instructions](https://www.nextflow.io/docs/latest/getstarted.html#installation)), Docker ([installation instructions](https://docs.docker.com/get-docker/)) and git ([installation instructions](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)) installed.

With Nextflow, Docker and git installed, clone the csgenetics_scrnaseq repository to a specified directory and change into that directory.

```bash
git clone https://github.com/csgenetics/csgenetics_scrnaseq.git $HOME/analysis && cd $HOME/analysis
```

The pipeline is then run using the Nextflow executable and the 'docker' profile.

```bash
nextflow run main.nf -profile docker --input_csv $HOME/analysis/input_csv/input_csv.csv
```

For a full list of the configurable parameters that can be can be supplied to the pipeline
and other options for configuration see the [usage docs](docs/usage.md).
<div style="text-align: right"><a href="#cs-genetics-scrna-seq-pipeline">top</a></div>

### Dockerized

The pipeline is run in a Docker container that has Nextflow installed. Within the container, the pipeline uses dedicated Docker containers for each of the executed tasks. This saves having to have Nextflow installed.

To run the pipeline dockerized, you must have Docker ([installation instructions](https://docs.docker.com/get-docker/)) and git ([installation instructions](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)) installed.

With Docker and git installed, clone the csgenetics_scrnaseq repository to a specified directory and change into that directory.

```bash
git clone https://github.com/csgenetics/csgenetics_scrnaseq.git $HOME/analysis && cd $HOME/analysis
```

The pipeline is then run by starting a docker container and supplying
it with commandline arguments. E.g.

```bash
docker run -it --rm -e USER=$USER \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v $HOME/analysis:$HOME/analysis \
  public.ecr.aws/csgenetics/csgenetics_scrnaseq_base:0.0.1 \ 
  nextflow run main.nf -profile docker -w $HOME/analysis/work/ --outdir $HOME/analysis/results/ --input_csv $HOME/analysis/input_csv/input_csv.csv
```

where:

- `-it --rm ` runs the container interactively and cleans up after exiting.

- `-e USER=$USER` makes your USER environmental variable visible instide the container.

- `-v /var/run/docker.sock:/var/run/docker.sock` mounts `/var/run/docker.sock` at `/var/run/docker.sock` enabling Docker to spawn sister containers.

- `-v $HOME/analysis:$HOME/analysis` mounts `$HOME/analysis` at `$HOME/analysis` setting up a persistent working directory for the pipeline run.

- `nextflow run main.nf ...` is the command for launching the pipeline including commandline-suopplied arguments (...)

For a full list of the configurable parameters that can be can be supplied to the pipeline
and other options for configuration see the [usage docs](docs/usage.md).
<div style="text-align: right"><a href="#cs-genetics-scrna-seq-pipeline">top</a></div>

## Specifying input sequencing files
The sequencing files to be analysed are specified using an input csv file.

An example template can be found [here](input_csv/template.csv).

The header row must be present and the full paths to files should be given.

The `fastq_1` should contain the sequencing data that will be mapped to the genome. `fastq_2` should contain the CS Genetics barcode.

The full path to the input csv should be supplied to the pipeline using the `--input_csv` flag.

E.g.
```bash
nextflow run main.nf -profile docker --input_csv <path/to/csv_dir/input_csv.csv>
```
<div style="text-align: right"><a href="#cs-genetics-scrna-seq-pipeline">top</a></div>

## Testing the pipeline
To test that your environment is setup correctly the pipeline can be run using the test profile:

```bash
nextflow run main.nf -profile test
```

The test proile will run using a set of remotely hosted resources. By default the work and results directories will be created in the current working directory at `./work` and `./results`, respectively.

For a full list of the configurable parameters that can be can be supplied to the pipeline
and other options for configuration see the [usage docs](docs/usage.md).
<div style="text-align: right"><a href="#cs-genetics-scrna-seq-pipeline">top</a></div>

## Launching the pipeline directly from the csgenetic/csgenetics_scrnaseq Github repo
In the above examples, the csgenetics_scrnaseq git repostory was cloned locally
and the pipeline was launched specifying the main.nf Nextflow script.

Alternatively the pipeline can be launched directly from the GitHub repository specifying its qualified name: `csgenetics/csgenetics_scrnaseq`.

See the [Nextflow documentation on Pipeline sharing](https://www.nextflow.io/docs/latest/sharing.html) for further details.
<div style="text-align: right"><a href="#cs-genetics-scrna-seq-pipeline">top</a></div>
